app/src/main/java/com/sunasterisk/englishflashcard/ui/addenglish/AddEnglishFragment.kt

```diff
@@ -1,13 +1,90 @@
+import com.sunasterisk.englishflashcard.ui.learnandsearch.TopicAdapter
+import com.sunasterisk.englishflashcard.ui.learnandsearch.TopicContract
+import com.sunasterisk.englishflashcard.ui.learnandsearch.TopicPresenter
+import com.sunasterisk.englishflashcard.ui.toast
+import kotlinx.android.synthetic.main.dialog_custom.view.*
+import kotlinx.android.synthetic.main.fragment_add_english.*
+import java.lang.Exception
 
-class AddEnglishFragment : BaseFragment() {
+class AddEnglishFragment : BaseFragment(), TopicContract.View, AddEnglishContract.View {
+
+    private var topicAdapter = TopicAdapter()
+    private var topicPresenter: TopicContract.Presenter? = null
+    private var addEnglishPresenter: AddEnglishContract.Presenter? = null
 
     override val layoutResource get() = R.layout.fragment_add_english
 
     override fun initComponents() {
+        initDataTopics()
+        addTopic()
+    }
+
+    private fun initDataTopics() {
+        context?.let {
+            val topicsDao = TopicsDAOlmpl.getInstance(it)
+            val localRepository: TopicsRepository = TopicsRepository.getInstance(
+                local = TopicsLocalDataSource.getInstance(topicsDao)
+            )
+            topicPresenter = TopicPresenter(this, localRepository)
+            topicPresenter?.getAllTopics()
+        }
+        recyclerViewTopics.adapter = topicAdapter
+    }
+
+    @SuppressLint("InflateParams")
+    fun addTopic() {
+        buttonAddTopic.setOnClickListener {
+            val builder = AlertDialog.Builder(context, R.style.CustomAlertDialog)
+            val view = layoutInflater.inflate(R.layout.dialog_custom, null)
+            builder.setView(view)
+            val alertDialog: AlertDialog = builder.create()
+            alertDialog.show()
+            val topicsDao = TopicsDAOlmpl.getInstance(context!!)
+            val localRepository: TopicsRepository = TopicsRepository.getInstance(
+                local = TopicsLocalDataSource.getInstance(topicsDao)
+            )
+            addEnglishPresenter = AddEnglishPresenter(this, localRepository)
+            view.buttonAdd.setOnClickListener {
+                addEnglishPresenter!!.addTopic(
+                    Topic(
+                        0,
+                        view.editTextNameTopic.text.toString(),
+                        view.editTextLinkImage.text.toString()
+                    )
+                )
+                alertDialog.dismiss()
+                initDataTopics()
+            }
+
+            view.buttonExit.setOnClickListener {
+                alertDialog.dismiss()
+            }
+        }
+    }
+
+    override fun showTopics(topics: List<Topic>) {
+        topicAdapter.updateData(topics)
+    }
+
+    override fun checkAddSuccess() {
+        context?.toast(getString(R.string.messeage_insert_success))
+    }
+
+    override fun checkAddFailed() {
+        context?.toast(getString(R.string.messeage_insert_failed))
+    }
```

Comment:
gộp vào, View hạn chế biết logic 
```suggestion
   override fun notifyMessage(messageResId: Int) {
        context?.toast(messageResId)
    }
```